File: native-modules/react-native-image-crop-picker/android/src/main/java/com/margelo/nitro/reactnativeimagecroppicker/ReactNativeImageCropPicker.kt
package com.margelo.nitro.reactnativeimagecroppicker

import android.os.Handler
import android.os.Looper
import androidx.activity.ComponentActivity
import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.Promise

@DoNotStrip
class ReactNativeImageCropPicker : HybridReactNativeImageCropPickerSpec() {
  private val mainHandler = Handler(Looper.getMainLooper())

  // Main thread only.
  private var activeSession: ImageCropPickerSession? = null

  override fun openPicker(options: ImageCropPickerOptions): Promise<PickedImage> =
    startSession(
      ImageCropPickerSession.Mode.Picker,
      ImageCropPickerConfig.from(options, forceCropping = false),
    )

  override fun openCropper(path: String, options: ImageCropPickerOptions): Promise<PickedImage> =
    startSession(
      ImageCropPickerSession.Mode.Cropper(path),
      ImageCropPickerConfig.from(options, forceCropping = true),
    )

  override fun clean(): Promise<Unit> =
    Promise.parallel {
      val context = NitroModules.applicationContext ?: throw ImageCropPickerException.cleanupFailed()
      ImageCropPickerImageProcessor.cleanTemporaryDirectory(context)
    }

  override fun cleanSingle(path: String): Promise<Unit> =
    Promise.parallel { ImageCropPickerImageProcessor.removeFile(path) }

  private fun startSession(
    mode: ImageCropPickerSession.Mode,
    config: ImageCropPickerConfig,
  ): Promise<PickedImage> {
    val promise = Promise<PickedImage>()
    mainHandler.post {
      activeSession?.let { session ->
        if (session.isActive) {
          promise.reject(ImageCropPickerException.inProgress())
          return@post
        }
        // Its activity went away without calling back; don't block new requests on it.
        session.abandon()
      }

      val activity = NitroModules.applicationContext?.currentActivity as? ComponentActivity
      if (activity == null || activity.isFinishing || activity.isDestroyed) {
        promise.reject(ImageCropPickerException.noActivity())
        return@post
      }

      val session = ImageCropPickerSession(activity, mode, config, promise) { finishedSession ->
        if (activeSession === finishedSession) {
          activeSession = null
        }
      }
      activeSession = session
      session.start()
    }
    return promise
  }
}

File: native-modules/react-native-image-crop-picker/ios/ReactNativeImageCropPicker.swift
import NitroModules
import UIKit

class ReactNativeImageCropPicker: HybridReactNativeImageCropPickerSpec {
  // Main thread only.
  private var activeSession: ImageCropPickerSession?

  func openPicker(options: ImageCropPickerOptions) throws -> Promise<PickedImage> {
    return startSession(mode: .picker, config: ImageCropPickerConfig(options))
  }

  func openCropper(path: String, options: ImageCropPickerOptions) throws -> Promise<PickedImage> {
    return startSession(
      mode: .cropper(path: path),
      config: ImageCropPickerConfig(options, forceCropping: true)
    )
  }

  func clean() throws -> Promise<Void> {
    return Promise.parallel {
      try ImageCropPickerImageProcessor.cleanTemporaryDirectory()
    }
  }

  func cleanSingle(path: String) throws -> Promise<Void> {
    return Promise.parallel {
      try ImageCropPickerImageProcessor.removeFile(atPath: path)
    }
  }

  private func startSession(
    mode: ImageCropPickerSession.Mode,
    config: ImageCropPickerConfig
  ) -> Promise<PickedImage> {
    let promise = Promise<PickedImage>()
    DispatchQueue.main.async {
      self.beginSession(mode: mode, config: config, promise: promise)
    }
    return promise
  }

  private func beginSession(
    mode: ImageCropPickerSession.Mode,
    config: ImageCropPickerConfig,
    promise: Promise<PickedImage>
  ) {
    if let activeSession {
      if activeSession.isActive {
        promise.reject(withError: ImageCropPickerError.inProgress)
        return
      }
      // Its UI went away without calling back; don't block new requests on it.
      activeSession.abandon()
    }

    let session = ImageCropPickerSession(mode: mode, config: config, promise: promise) {
      [weak self] finishedSession in
      if self?.activeSession === finishedSession {
        self?.activeSession = nil
      }
    }
    activeSession = session
    session.start()
  }
}
